posts/2024-09-05-impl-trait-capture-rules.md

```diff
@@ -0,0 +1,248 @@
+---
+layout: post
+title: "Changes to `impl Trait` in Rust 2024"
+author: Niko Matsakis
+team: the language team <https://www.rust-lang.org/governance/teams/lang>
+---
+The default way `impl Trait` works in return position is changing in Rust 2024. These changes are meant to simplify `impl Trait` to better match what people want most of the time. We're also adding a flexible syntax that gives you full control when you need it.
+
+## TL;DR
+
+Starting in Rust 2024, we are changing the rules for when a generic parameter can be used in the hidden type of a return-position `impl Trait`:
+
+* a new default that the hidden types for a return-position `impl Trait` can use **any** generic parameter in scope, instead of only types (applicable only in Rust 2024);
+* a syntax to declare explicitly what types may be used (usable in any edition).
+
+The new explicit syntax is called a "use bound": `impl Trait + use<'x, T>`, for example, would indicate that the hidden type is allowed to use `'x` and `T` (but not any other generic parameters in scope).
+
+Read on for the details!
+
+## Background: return-position `impl Trait`
+
+This blog post concerns *return-position `impl Trait`*, such as the following example:
+
+```rust
+fn process_data(
+    data: &[Datum]
+) -> impl Iterator<Item = ProcessedDatum> {
+    data
+        .iter()
+        .map(|datum| datum.process())
+}
+```
+
+The use of `-> impl Iterator` in return position here means that the function returns "some kind of iterator". The actual type will be determined by the compiler based on the function body. It is called the "hidden type" because callers do not get to know exactly what it is; they have to code against the `Iterator` trait. However, at code generation time, the compiler will generate code based on the actual precise type, which ensures that callers are fully optimized.
+
+Although callers don't know the exact type, they do need to know that it will continue to borrow the `datums` argument so that they can ensure that the `datums` reference remains valid while iteration occurs. Further, callers must be able to figure this out based solely on the type signature, without looking at the function body.
```

Comment:
Pluralization may have been partial, there are a handful of `datums` left.

```suggestion
Although callers don't know the exact type, they do need to know that it will continue to borrow the `data` argument so that they can ensure that the `data` reference remains valid while iteration occurs. Further, callers must be able to figure this out based solely on the type signature, without looking at the function body.
```